qa/L0_model_update/instance_update_test.py

```diff
@@ -0,0 +1,310 @@
+# are met:
+#  * Redistributions of source code must retain the above copyright
+#    notice, this list of conditions and the following disclaimer.
+#  * Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in the
+#    documentation and/or other materials provided with the distribution.
+#  * Neither the name of NVIDIA CORPORATION nor the names of its
+#    contributors may be used to endorse or promote products derived
+#    from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import unittest
+import time
+import concurrent.futures
+import numpy as np
+import tritonclient.grpc as grpcclient
+from tritonclient.utils import InferenceServerException
+from models.instance_init_del.util import *
+
+
+class TestInstanceUpdate(unittest.TestCase):
+
+    def setUp(self):
+        # Reset counters
+        reset_initialize_count()
+        reset_finalize_count()
+        # Reset batching
+        disable_batching()
+        # Reset delays
+        set_initialize_delay(0)
+        set_infer_delay(0)
+        # Initialize client
+        self.__triton = grpcclient.InferenceServerClient("localhost:8001",
+                                                         verbose=True)
+
+    def __get_inputs(self, shape=(2,)):
+        inputs = [grpcclient.InferInput("INPUT0", shape, "FP32")]
+        inputs[0].set_data_from_numpy(np.ones(shape, dtype=np.float32))
+        return inputs
+
+    def test_add_rm_add(self):
+        # Load model
+        update_instance_group("{\ncount: 3\nkind: KIND_CPU\n}")
+        self.__triton.load_model("instance_init_del")
+        self.assertEqual(get_initialize_count(), 3)
+        self.assertEqual(get_finalize_count(), 0)
+        self.__triton.infer("instance_init_del", self.__get_inputs((8,)))
+        # Add 1 instance
+        update_instance_group("{\ncount: 4\nkind: KIND_CPU\n}")
+        self.__triton.load_model("instance_init_del")
+        self.assertEqual(get_initialize_count(), 4)
+        self.assertEqual(get_finalize_count(), 0)
+        self.__triton.infer("instance_init_del", self.__get_inputs((2,)))
```

Comment:
Are we testing something specific here? We don't seem to be validating the results. I believe a better test would be to run 4 (number of active instance) async infer calls and check the results are available within specified time interval. (the inference occurred concurrently)

Reply:
This can be done within a utility function that takes the number of instance counts.

Reply:
> I believe a better test would be to run 4 (number of active instance) async infer calls and check the results are available within specified time interval. (the inference occurred concurrently)

I do not think concurrent update will happen, because the `model_repository_manager` will serialize update (load/unload) to a model, which at best will only make the updates to be out-of-order.

> Are we testing something specific here?

Yes. This is to test if data structures can return to its original state by `original -> [add 1 item] original + 1 -> [remove 1 item] original -> [add 1 item] original + 1`. This is just to see if the implementation is well-formed (not to stress anything), because if something is missing, this simple case could break.

Reply:
I meant the concurrent inference request and not concurrent updates.

Reply:
> utility function

Updated: [Refactor instance update test](https://github.com/triton-inference-server/server/pull/5635/commits/f012b0474ac4b1465e432e72f3c4f24cb5cb6b3c)

Reply:
> I meant the concurrent inference request and not concurrent updates.

Updated: [Add instance update with parallel infer](https://github.com/triton-inference-server/server/pull/5635/commits/27b9f9a0217b3a72df71a490541fd392e67d7b1b)